coucharchive

```diff
@@ -240,34 +241,53 @@ class CouchDBInstance(object):
         self.thread = None
 
 
-# Create a throttler object that can be safely shared between processes
-manager = multiprocessing.Manager()
-throttler = manager.Namespace()
-throttler.lock = manager.Lock()
-throttler.delay = 0
+class ReplicationControl(object):
+    def __init__(self, total_replications):
+        self.total_replications = total_replications
+        self.running_replications = multiprocessing.Manager().Value('i', 0)
+        self.completed_replications = multiprocessing.Manager().Value('i', 0)
+        self.last_successes = multiprocessing.Manager().list()
+        self.last_errors = multiprocessing.Manager().list()
 
+    def report_success(self):
+        self.last_successes.append((int(time.time()),
+                                    self.running_replications.value))
+        self.last_successes = self._recent_events_only(self.last_successes)
 
-def get_throttler_delay():
-    global throttler
-    with throttler.lock:
-        return throttler.delay
+    def report_error(self):
+        self.last_errors.append((int(time.time()),
+                                 self.running_replications.value))
+        self.last_errors = self._recent_events_only(self.last_errors)
```

Comment:
Hey!
I think there is a small issue here, as the `self.last_errors` list is clean up only when new error are reported, i.e at line 276, values that are in the list can be really old.
May I suggest keeping the `recent_events_only` just before using the list, i.e at line 275? (same for `self.last_successes` list but success are reported way more often than failure)

Reply:
Genius!

Your proposal seems idea, can you fix and push to the branch?

Reply:
Hey!

So I try to do it the simpliest possible way:
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -268,11 +268,13 @@ class ReplicationControl(object):
         ideal_number = 4
 
         # ... then use 2 × the last successful value
+        self.last_successes = self._recent_events_only(self.last_successes)
         if self.last_successes:
             best_successful_number = max(e[1] for e in self.last_successes)
             ideal_number = 2 * best_successful_number
 
         # ... but if there were errors, take 0.9 × the lowest error value
+        self.last_errors = self._recent_events_only(self.last_errors)
         if self.last_errors:
             min_number_with_error = min(e[1] for e in self.last_errors)
             ideal_number = int(0.9 * min_number_with_error)
@@ -286,6 +288,7 @@ class ReplicationControl(object):
         # If there was 1 error within last 5 minutes, pause for 5 seconds,
         # if there were 2 errors within last 5 minutes, pause for 10 seconds,
         # if there were 3 errors within last 5 minutes, pause for 15 seconds...
+        self.last_errors = self._recent_events_only(self.last_errors)
         wait = 5 * len(self.last_errors)
         return min(60, wait)  # stay within [5 s, 60 s]
```
**But**, when I do that, I have the following error that is raised in the `report_success()` function which I did not change:
```
  File "./coucharchive", line 507, in replicate_one_database
    control.report_success()
  File "./coucharchive", line 254, in report_success
    self.running_replications.value))
  File "<string>", line 2, in append
  File "/usr/lib64/python3.7/multiprocessing/managers.py", line 815, in _callmethod
    self._connect()
  File "/usr/lib64/python3.7/multiprocessing/managers.py", line 802, in _connect
    conn = self._Client(self._token.address, authkey=self._authkey)
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 492, in Client
    c = SocketClient(address)
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 619, in SocketClient
    s.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
INFO: chatterton_common_billables_ru: done
Process ForkPoolWorker-10:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/managers.py", line 811, in _callmethod
    conn = self._tls.connection
AttributeError: 'ForkAwareLocal' object has no attribute 'connection'
```
Here is the only popular discussion I find in the Internet, **but for us the case is clearly more subtle**:
https://stackoverflow.com/questions/25455462/share-list-between-process-in-python-server

So, one solution I find is to read lists `self.last_successes ` and `self.last_errors` inside `ideal_*()` functions, but not write inside them. 
It's not ideal, but finding another solution seems to me to be a long task.
Here is the diff pushed in a `FIXUP` commit:
```diff
 --- a/coucharchive
+++ b/coucharchive
@@ -268,13 +268,15 @@ class ReplicationControl(object):
         ideal_number = 4
 
         # ... then use 2 × the last successful value
-        if self.last_successes:
-            best_successful_number = max(e[1] for e in self.last_successes)
+        last_successes = self._recent_events_only(self.last_successes)
+        if last_successes:
+            best_successful_number = max(e[1] for e in last_successes)
             ideal_number = 2 * best_successful_number
 
         # ... but if there were errors, take 0.9 × the lowest error value
-        if self.last_errors:
-            min_number_with_error = min(e[1] for e in self.last_errors)
+        last_errors = self._recent_events_only(self.last_errors)
+        if last_errors:
+            min_number_with_error = min(e[1] for e in last_errors)
             ideal_number = int(0.9 * min_number_with_error)
 
         # ... and stay within [1, MAX_NUMBER_OF_WORKERS]
@@ -286,7 +288,8 @@ class ReplicationControl(object):
         # If there was 1 error within last 5 minutes, pause for 5 seconds,
         # if there were 2 errors within last 5 minutes, pause for 10 seconds,
         # if there were 3 errors within last 5 minutes, pause for 15 seconds...
-        wait = 5 * len(self.last_errors)
+        last_errors = self._recent_events_only(self.last_errors)
+        wait = 5 * len(last_errors)
         return min(60, wait)  # stay within [5 s, 60 s]
```
What do you think?
For info I launch a test on my testing servers to see what's what.

Reply:
> But, when I do that, I have the following error that is raised in the report_success() function which I did not change:

Hehe, there's a trick here! You can simply replace the array with another: it's not a `[]`, it's a `multiprocessing.Manager().list()`. If you do `last_successes = ...`, it won't be a `multiprocessing.Manager().list()` anymore, so it breaks when accessed concurrently.

In order to replace its values, without destroying/recreating it, you can do `last_successes[:] = ...`.

I replaced your fixup commit with the following:
```diff
--- a/coucharchive
+++ b/coucharchive
@@ -252,18 +252,21 @@ class ReplicationControl(object):
     def report_success(self):
         self.last_successes.append((int(time.time()),
                                     self.running_replications.value))
-        self.last_successes = self._recent_events_only(self.last_successes)
 
     def report_error(self):
         self.last_errors.append((int(time.time()),
                                  self.running_replications.value))
-        self.last_errors = self._recent_events_only(self.last_errors)
 
-    def _recent_events_only(self, events):
+    def _drop_old_events(self):
         five_min_ago = time.time() - 5 * 60
-        return [e for e in events if e[0] > five_min_ago]
+        self.last_successes[:] = [e for e in self.last_successes
+                                  if e[0] > five_min_ago]
+        self.last_errors[:] = [e for e in self.last_errors
+                               if e[0] > five_min_ago]
 
     def ideal_number_of_replications(self):
+        self._drop_old_events()
+
         # Start with 4 concurrent replications...
         ideal_number = 4
 
@@ -283,6 +286,8 @@ class ReplicationControl(object):
         return ideal_number
 
     def ideal_sleep_value(self):
+        self._drop_old_events()
+
         # If there was 1 error within last 5 minutes, pause for 5 seconds,
         # if there were 2 errors within last 5 minutes, pause for 10 seconds,
         # if there were 3 errors within last 5 minutes, pause for 15 seconds...
```